devtools/src/main/scala/kalix/devtools/impl/DockerComposeUtils.scala

```diff
@@ -35,22 +37,62 @@ object DockerComposeUtils {
 
       throw new IllegalArgumentException(s"File '$file' does not exist. $extraMsg")
     }
+  }
 
-  def start(file: String, userFunctionPort: Int): Unit = {
-
-    execIfFileExists(file) {
-      Process(s"docker-compose -f $file up", None, "USER_FUNCTION_PORT" -> userFunctionPort.toString).run
+  // read the file once and cache the lines
+  // we will need to iterate over it more than once
+  private lazy val lines: Seq[String] = {
+    if (Files.exists(Paths.get(file))) {
+      val src = Source.fromFile(file)
+      try {
+        src.getLines.toList
+      } finally {
+        src.close()
+      }
+    } else {
+      Seq.empty
     }
+  }
+
+  def start(): Unit = {
+    execIfFileExists {
+      Process(s"docker-compose -f $file up", None).run
 
-    // shutdown hook to down containers when jvm exits
-    sys.addShutdownHook {
-      execIfFileExists(file) {
-        stop(file)
+      // shutdown hook to down containers when jvm exits
+      sys.addShutdownHook {
+        execIfFileExists {
+          stop()
+        }
       }
     }
   }
 
-  private def stop(file: String): Unit = {
+  private def stop(): Unit = {
     s"docker compose -f $file stop".!
   }
+
+  def readUserFunctionPort: Int = {
+    envVar
+      .get("USER_FUNCTION_PORT")
+      .map(_.toInt)
+      .orElse(readUserFunctionPortFromFile)
+      .getOrElse(8080)
+  }
+
+  private val ExtractPort = """\D+:*(\d+)""".r
+  private def readUserFunctionPortFromFile: Option[Int] = {
+    lines
+      .find(_.trim.startsWith("USER_FUNCTION_PORT"))
+      // TODO: remove this hack
+      // regex doesn't match if the line contains a `}`
+      // strange enough I tested it online and in other programs and it works
+      // I guess this is something related to Java regex. Leaving as is for now
+      .map(_.trim.replace("}", ""))
+      .collect { case ExtractPort(port) => port.toInt }
+  }
+
+  def readServicePortMappings: Seq[String] =
+    lines
+      .map(_.trim)
+      .filter(_.startsWith("-Dkalix.dev-mode.service-port-mappings"))
```

Comment:
This is not sufficient. 
One can configure their docker with:

`-Dsomething=bla -Dkalix.dev-mode.service-port-mappings.foo=9001` 

And then we won't find the service-port-mappings because line is not starting with it.

Need to make it more robust.